### PR TITLE
[8.x] Cursor pagination fixes

### LINKED ISSUE #37400: Edge case breaking Cursor Pagination
- Laravel Version: 8.41.0
- PHP Version: 8.0.6
- Database Driver & Version: mysql  Ver 14.14 Distrib 5.7.33-36

### Description:

It seems that the new Cursor Pagination from #37216 does not work well when specifying the table used in `orderBy` / `latest` (when working with inner join / ambiguous columns).

Edit: Actually, it seems like this is related to pivots

### Steps To Reproduce:

I've got the following relationship on my `User` model, an user can hide other users:

```php
public function hidden()
{
    return $this
        ->belongsToMany(User::class, 'hidden_profiles', 'user_id', 'hidden_id')
        ->withTimestamps();
}
```

What I would like to do is displaying that list of hidden users, ordered by the latest added.
Simply done like that:

```php
$user->hidden()->latest('created_at')->cursorPaginate();
```

However, this does not work because of ambiguous columns, both `users` and `hidden_profiles` have `created_at` columns, so we have to give the proper table:

```php
$user->hidden()->latest('hidden_profiles.created_at')->cursorPaginate();
```

This works, but sadly once you give him a cursor (browsing next page), it will break with the following error:

```
Only arrays and objects are supported when cursor paginating items.
```

I will try to work on a patch or at least a repo / failing test to reproduce this easily as soon as I can.

@paras-malhotra Is this a limitation by design? Any idea where I should start?

### PR DESCRIPTION
**Pull request by driesvints**
This PR fixes ~two~ three things:

- Proper column distribution for cursor base pagination when working with pivots (probably just an oversight from https://github.com/laravel/framework/pull/37216
- Allow usage of aliased columns with order by statements when using cursor based pivots
- Proper attribute retrieval for cursors when combining them with pivot models

Before the second fix, the query builder would generate the following invalid query:

```sql
SELECT
    *,
    `friends`.`created_at` AS `friended_at`,
    `users`.*,
    `friends`.`user_id` AS `pivot_user_id`,
    `friends`.`friend_id` AS `pivot_friend_id`,
    `friends`.`created_at` AS `pivot_created_at`,
    `friends`.`updated_at` AS `pivot_updated_at`
FROM
    `users`
    INNER JOIN `friends` ON `users`.`id` = `friends`.`friend_id`
WHERE
    `friends`.`user_id` = ?
    AND `friended_at` < ?
ORDER BY
    `friended_at` DESC
LIMIT 6;
```

Now it will generate:

```sql
SELECT
    *,
    `friends`.`created_at` AS `friended_at`,
    `users`.*,
    `friends`.`user_id` AS `pivot_user_id`,
    `friends`.`friend_id` AS `pivot_friend_id`,
    `friends`.`created_at` AS `pivot_created_at`,
    `friends`.`updated_at` AS `pivot_updated_at`
FROM
    `users`
    INNER JOIN `friends` ON `users`.`id` = `friends`.`friend_id`
WHERE
    `friends`.`user_id` = ?
    AND `friends`.`created_at` < ?
ORDER BY
    `friended_at` DESC
LIMIT 6;
```

The reason for this is that sql engines need an actual column to sort and can't rely on an aliased column of a result set. By de-aliasing the given columns we circumvent this issue.

I don't think this contains anything breaking but I'm unsure about the usage to detect the column aliasing:

```php
stripos($column, ' as ') !== false
```

I saw this was used in the base query grammar class as well with the `wrap` method so I think this will be fine.

This PR will conflict with https://github.com/laravel/framework/pull/37762

Fixes https://github.com/laravel/framework/issues/37400

### CODE REVIEW

**Comment by driesvints** [comment]
@paras-malhotra would appreciate a review if you're willing.

**Comment by kamui545** [comment]
@driesvints Thank you for your hard work, really appreciate it.

Hydration seems to be working properly for the pivot now!

Sadly, I am still having weird behaviour when browsing the "next" page.

On my playground repo, if you're browsing the next page on the `cursorPaginate()` without alias, you will have the `Only arrays and objects are supported when cursor paginating items` error.

Using an alias seems to fix the issue on the playground, however, I tried in my own app and I get the same error even with the alias... I've got no idea what's going on, the code of the playground is based from my app and should be similar... I will keep you updated once I figured out... but for now im banging my head against the wall

**Comment by driesvints** [comment]
I'll need to check in on that on Thursday. I also see it failing but also can't figure it out atm.

**Comment by kamui545** [comment]
Ok so after spending some time debugging, the alias is now working properly on my real app as well. I was sorting on a non-unique column, my records all had the same timestamps and the next page returned no results, failing to build the previous cursor because `$this->items->first()` was `null`...

Stupid mistake, just like I did on the playground app: it should be sorting on `id` instead of `created_at` to make sure it's unique...

The app is still acting weird when not using an alias though. Eg:

```php
$friends = $this->user()
  ->friends()
  ->latest('friends.id')
  ->cursorPaginate($this->limit);
```

I am not sure if there is an easy way to fix this.

Maybe creating an alias automatically when a dot is used would do the trick. Just need to make sure it does not collide.

I will investigate further this week.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -844,11 +844,13 @@ public function cursorPaginate($perPage = null, $columns = ['*'], $cursorName =
 
         $parameters = $orders->pluck('column')->toArray();
 
+        $originals = $this->reverseColumnAliasingForCursorPagination($parameters);
+
         if (! is_null($cursor)) {
             if (count($parameters) === 1) {
-                $this->where($column = $parameters[0], $comparisonOperator, $cursor->parameter($column));
+                $this->where($originals[0], $comparisonOperator, $cursor->parameter($parameters[0]));
```
Comment (rafael-neris) [inline src/Illuminate/Database/Eloquent/Builder.php:851]: Heyy @driesvints, what do you think about use array_shift instead then `$originals[0]` and `$parameters[0]`?

Comment (rafael-neris) [inline src/Illuminate/Database/Eloquent/Builder.php:851]: `$this->where(array_shift($originals), $comparisonOperator, $cursor->parameter(array_shift($parameters)));`

**Comment by paras-malhotra** [comment]
@driesvints I went through the code. I think this PR looks good to go. Definitely seems to have been an oversight on my part to not hydrate the pivot relations and not fetch the original columns for the where clause.

I guess #37762 changed the where clauses entirely but this logic would still need to be applied to that PR.

File: src/Illuminate/Database/Eloquent/Builder.php (hunk 1 above)
Comment (driesvints) [reply src/Illuminate/Database/Eloquent/Builder.php:851]: Not necessary imo.

**Comment by driesvints** [comment]
@kamui545 think I got a fix in for the issue you were experiencing. The app you shared is now fully working with this PR. Can you try to verify that? 

@paras-malhotra we pushed one more commit to fix the above. It basically will retrieve attributes on a pivot relationship when the table matches with the first part of the dot notation of a parameter name. I'll try some more use case with other relationship queries tomorrow. But this will essentially allow to filter on non-unique column names as long as the table is prefixed.

**Comment by kamui545** [comment]
@driesvints Thank you again! It seems to fix ordering on ambiguous columns.

However, I noticed another issue: #37999

**Pull request merged**